File: genai-engine/ui/src/components/traces/components/add-to-dataset/utils/transformExecutor.ts
import { NestedSpanWithMetricsResponse } from "@/lib/api-client/api-client";
import { TransformDefinition, Column } from "../form/shared";

// Converts value to string: primitives as-is, objects/arrays via JSON.stringify
function stringifyValue(value: any): string {
  if (value === null || value === undefined) return "";
  if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
    return String(value);
  }
  
  try {
    return JSON.stringify(value);
  } catch (error) {
    return String(value);
  }
}

// Executes transform on spans, returns extracted columns. Uses first match if multiple spans found.
export function executeTransform(
  spans: NestedSpanWithMetricsResponse[],
  transform: TransformDefinition
): Column[] {
  const columns: Column[] = [];

  for (const colDef of transform.columns) {
    const matchingSpans = spans.filter((span) => span.span_name === colDef.span_name);

    if (matchingSpans.length === 0) {
      columns.push({
        name: colDef.column_name,
        value: stringifyValue(colDef.fallback ?? ""),
        path: `${colDef.span_name}.${colDef.attribute_path}`,
        span_name: colDef.span_name,
        attribute_path: colDef.attribute_path,
        matchCount: 0,
      });
    } else {
      const span = matchingSpans[0];
      const value = getNestedValue(span.raw_data, colDef.attribute_path);

      columns.push({
        name: colDef.column_name,
        value: value !== undefined && value !== null 
          ? stringifyValue(value) 
          : stringifyValue(colDef.fallback ?? ""),
        path: `${span.span_name}.${colDef.attribute_path}`,
        span_name: colDef.span_name,
        attribute_path: colDef.attribute_path,
        matchCount: matchingSpans.length,
        selectedSpanId: span.span_id,
        allMatches: matchingSpans.length > 1 ? matchingSpans.map(s => ({
          span_id: s.span_id,
          span_name: s.span_name || "unknown",
          extractedValue: stringifyValue(getNestedValue(s.raw_data, colDef.attribute_path)),
        })) : undefined,
      });
    }
  }

  return columns;
}

// Extracts value from nested object using dot-notation path
function getNestedValue(obj: any, path: string): any {
  if (!obj || !path) return undefined;

  const keys = path.split(".");
  let current = obj;

  for (const key of keys) {
    if (current && key in current) {
      current = current[key];
    } else {
      return undefined;
    }
  }

  return current;
}

File: genai-engine/src/utils/transform_executor.py
"""Transform execution utilities for dataset transforms.

This module provides functionality to execute transforms against trace spans,
matching the behavior of the frontend transform executor in
genai-engine/ui/src/components/traces/components/add-to-dataset/utils/transformExecutor.ts
"""

import json
from typing import Any, List

from arthur_common.models.response_schemas import (
    NestedSpanWithMetricsResponse,
    TraceResponse,
)

from schemas.request_schemas import TraceTransformDefinition
from schemas.response_schemas import (
    TransformExtractionResponseList,
    TransformExtractionResponseVariable,
)
from utils.trace import get_nested_value


def stringify_value(value: Any) -> str:
    """Convert value to string: primitives as-is, objects/arrays via JSON.stringify.

    Args:
        value: The value to convert to string

    Returns:
        String representation of the value
    """
    if value is None:
        return ""
    if isinstance(value, (str, int, float, bool)):
        return str(value)

    try:
        return json.dumps(value)
    except Exception:
        return str(value)


def _flatten_spans(
    span: NestedSpanWithMetricsResponse,
) -> List[NestedSpanWithMetricsResponse]:
    """Flatten nested span structure into a flat list.

    Args:
        span: The root span with potential children

    Returns:
        Flat list of all spans in the tree
    """
    result = [span]
    if span.children:
        for child in span.children:
            result.extend(_flatten_spans(child))
    return result


def execute_transform(
    trace: TraceResponse,
    transform_definition: TraceTransformDefinition,
) -> TransformExtractionResponseList:
    """Execute transform on a trace, returns raw extracted values.

    Base extraction function that returns raw variable names and values without
    formatting them into specific structures.

    Uses first match if multiple spans found. This matches the behavior of the
    frontend executeTransform function.

    Args:
        trace: TraceResponse object containing root_spans
        transform_definition: TraceTransformDefinition object containing variables to extract

    Returns:
        TransformExtractionResponseList containing list of variable names and values
    """
    # Flatten the nested span structure into a flat list
    flat_spans: List[NestedSpanWithMetricsResponse] = []
    for span in trace.root_spans:
        flat_spans.extend(_flatten_spans(span))

    variables = []

    # Iterate through variable definitions
    for var_def in transform_definition.variables:
        span_name = var_def.span_name
        variable_name = var_def.variable_name
        attribute_path = var_def.attribute_path
        fallback = var_def.fallback

        # Find matching spans by span_name
        matching_spans = [span for span in flat_spans if span.span_name == span_name]

        if not matching_spans:
            # No matching span found, use fallback
            value = fallback if fallback is not None else ""
        else:
            # Use the first matching span
            span = matching_spans[0]
            value = get_nested_value(span.raw_data, attribute_path)

            # Use value if found, otherwise use fallback
            if value is None:
                value = fallback if fallback is not None else ""

        variables.append(
            TransformExtractionResponseVariable(
                variable_name=variable_name,
                value=stringify_value(value),
            ),
        )

    return TransformExtractionResponseList(variables=variables)
